dcpy/utils/postgres.py

```diff
@@ -8,41 +8,131 @@
 from typing import Optional
 
 
-BUILD_ENGINE_RAW = os.environ["BUILD_ENGINE"]
-build_engine = create_engine(BUILD_ENGINE_RAW)
+def generate_engine_uri(server_url: str, database: str, schema: Optional[str] = None):
+    schemas = "public" if (not schema or schema == "public") else f"{schema},public"
+    options = f"?options=--search_path%3D{schemas}"
+    return server_url + "/" + database + options
 
 
-def execute_query(query: str, placeholders: Optional[dict] = None) -> None:
-    if placeholders is None:
-        placeholders = {}
-    with build_engine.connect() as connection:
-        connection.execute(statement=text(query), parameters=placeholders)
-
-
-def execute_select_query(
-    query: str, placeholders: Optional[dict] = None
-) -> pd.DataFrame:
-    if placeholders is None:
-        placeholders = {}
-    with build_engine.connect() as sql_conn:
-        select_records = pd.read_sql(sql=text(query), con=sql_conn, params=placeholders)
-    return select_records
-
-
-def execute_file_via_shell(build_engine: str, path: Path):
+def execute_file_via_shell(engine_uri: str, path: Path):
     """Execute .sql script at given path."""
-    cmd = f"psql {build_engine} -v ON_ERROR_STOP=1 -f {path}"
+    cmd = f"psql {engine_uri} -v ON_ERROR_STOP=1 -f {path}"
     if os.system(cmd) != 0:
         raise Exception(f"{path} has errors!")
 
 
-def execute_via_shell(build_engine: str, sql_statement):
+def execute_query_via_shell(engine_uri: str, sql_statement):
     """Execute sql via psql shell."""
-    cmd = f"psql {build_engine} -v ON_ERROR_STOP=1 {sql_statement}"
+    cmd = f"psql {engine_uri} -v ON_ERROR_STOP=1 {sql_statement}"
     if os.system(cmd) != 0:
         raise Exception(f"Command has errors! {cmd}")
 
 
+class PostgresClient:
+    def __init__(
+        self,
+        schema: Optional[str] = None,
+        database: str = os.environ["BUILD_ENGINE_DB"],
```

Comment:
So just a note... these defaults are resolved when the module is imported, not when the class is instantiated. So this will throw an exception in a product like facdb, where we just use the `postgres.insert_copy` function, unless we set `BUILD_ENGINE_DB` and `BUILD_ENGINE_SERVER` there as well

Reply:
Ah yeah this addresses my question just asked [here](https://github.com/NYCPlanning/data-engineering/pull/234/files#r1333285247)

Reply:
yea I'd rather not force setting envars when they won't be used.

per [my comment above](https://github.com/NYCPlanning/data-engineering/pull/234#discussion_r1331836745), this is the current approach in our `s3.client` class. but looks like every function in the module would instantiate a client, so maybe the approach is good to keep using there?

Reply:
@alexrichey see [41a0d8f](https://github.com/NYCPlanning/data-engineering/pull/234/commits/41a0d8f77038249f2abbc61b90732c1b387dd088)

my typing conundrum was because I was trying to pass along `database` rather than `self.database`